File: src/core/net/envelope.test.ts
import { describe, expect, it } from 'vitest'
import { readFixtureBytes } from './__fixtures__'
import { decodeGb18030 } from './encoding/gb18030'
import { parseNgaJson } from './envelope'
import { NgaError, extractServerError, isFakeError } from './errors'

const fixtureText = (name: Parameters<typeof readFixtureBytes>[0]) =>
  decodeGb18030(readFixtureBytes(name))

describe('isFakeError · 假错误白名单（API 文档 §0.7）', () => {
  it.each(['完毕', '没找到', '没有符合条件的结果', '今天已经签到', '找不到用户'])(
    '%s 视为成功',
    (message) => {
      expect(isFakeError(message)).toBe(true)
    },
  )

  it('白名单按子串匹配：发贴完毕 / 操作完毕', () => {
    expect(isFakeError('发贴完毕')).toBe(true)
    expect(isFakeError('操作完毕，正在跳转')).toBe(true)
  })

  it('真错误不在白名单里', () => {
    expect(isFakeError('找不到主题')).toBe(false)
    expect(isFakeError('未登录')).toBe(false)
    expect(isFakeError('您没有权限进行此操作')).toBe(false)
  })
})

describe('extractServerError', () => {
  it('无 code 时补 ?', () => {
    expect(extractServerError({ error: { '0': '未登录' } })).toEqual({ code: '?', message: '未登录' })
  })

  it('取 error.code', () => {
    expect(extractServerError({ error: { code: 403, '0': '找不到主题' } })).toEqual({
      code: 403,
      message: '找不到主题',
    })
  })

  it('多条信息用；连起来', () => {
    expect(extractServerError({ error: { '0': 'a', '1': 'b', '2': '' } })).toEqual({
      code: '?',
      message: 'a；b',
    })
  })

  it('没有 error 就是 null', () => {
    expect(extractServerError({ data: {}, time: 1 })).toBeNull()
    expect(extractServerError('not an object')).toBeNull()
  })
})

describe('parseNgaJson', () => {
  it('取出 data 与 time', () => {
    const envelope = parseNgaJson('{"data":{"0":"ok"},"time":1786111705}')
    expect(envelope.data).toEqual({ '0': 'ok' })
    expect(envelope.time).toBe(1786111705)
    expect(envelope.fakeError).toBeUndefined()
  })

  it('顶层既没有 data 也没有 error 时默认报解析错（可重试，链会换下一个组合）', () => {
    // 以前这里把顶层当 data，于是任何一个陌生 JSON 都成了「合法的空数据」，
    // 一路走到 UI 变成「这个版块还没有主题」（2026-08-13，「版块全空」排查）
    try {
      parseNgaJson('{"code":0,"msg":"","result":[]}', 'direct')
      expect.unreachable('应当抛错')
    } catch (error) {
      expect(error).toBeInstanceOf(NgaError)
      expect((error as NgaError).kind).toBe('parse')
      expect((error as NgaError).retryable).toBe(true)
      expect((error as NgaError).via).toBe('direct')
    }
  })

  it('调用方显式声明 bare 时，顶层才当 data', () => {
    const envelope = parseNgaJson('{"code":0,"msg":"","result":[]}', undefined, 'bare')
    expect(envelope.data).toEqual({ code: 0, msg: '', result: [] })
  })

  it('有 data 壳的响应不受 bare 影响', () => {
    expect(parseNgaJson('{"data":{"0":"ok"}}', undefined, 'bare').data).toEqual({ '0': 'ok' })
  })

  it('真错误抛 server 错误，且不重试', () => {
    const text = fixtureText('readThreadNotFound')
    expect(() => parseNgaJson(text)).toThrowError(NgaError)
    try {
      parseNgaJson(text, 'direct')
      expect.unreachable('应当抛错')
    } catch (error) {
      expect(error).toBeInstanceOf(NgaError)
      const ngaError = error as NgaError
      expect(ngaError.kind).toBe('server')
      expect(ngaError.message).toContain('找不到主题')
      expect(ngaError.retryable).toBe(false)
      expect(ngaError.via).toBe('direct')
    }
  })

  it('假错误当成功返回，但把错误信息留给调用方判空', () => {
    const envelope = parseNgaJson(fixtureText('ucpNotFound'))
    expect(envelope.fakeError?.message).toBe('找不到用户')
    expect(envelope.data).toBeUndefined()
  })

  it('解析失败抛 parse 错误，且可重试（解析失败 ≈ 被封）', () => {
    try {
      parseNgaJson('<html>你被封了</html>')
      expect.unreachable('应当抛错')
    } catch (error) {
      const ngaError = error as NgaError
      expect(ngaError.kind).toBe('parse')
      expect(ngaError.retryable).toBe(true)
    }
  })

  it('空响应也算 parse 错误', () => {
    expect(() => parseNgaJson('')).toThrowError(/响应为空/)
  })

  it('真实抓包：通知接口拿到 data', () => {
    const envelope = parseNgaJson(fixtureText('notiEmpty'))
    expect(envelope.data).toEqual({ '0': '' })
  })

  it('真实抓包：用户资料', () => {
    const envelope = parseNgaJson(fixtureText('ucpUser'))
    const user = (envelope.data as Record<string, Record<string, unknown>>)['0']!
    expect(user.uid).toBe(41417929)
    expect(user.username).toBe('BugenZhao')
  })

  it('真实抓包：主题列表（GBK + 未声明 charset 的那条）', () => {
    const envelope = parseNgaJson(fixtureText('threadList'))
    const data = envelope.data as Record<string, Record<string, Record<string, unknown>>>
    expect(data.__F!.name).toBe('原神')
    expect(Object.keys(data.__T!).length).toBeGreaterThan(10)
  })

  it('真实抓包：read.php 的 lite=js 前缀被剥掉后能解析', () => {
    const envelope = parseNgaJson(fixtureText('readThread'))
    const data = envelope.data as Record<string, unknown>
    expect(data.__R).toBeDefined()
  })
})

File: app/src/test/kotlin/com/chasel/ng2n/core/net/EnvelopeTest.kt
package com.chasel.ng2n.core.net

import com.chasel.ng2n.golden.Goldens
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class EnvelopeTest {

  private fun captureText(name: String): String =
    Goldens.load("envelope").first { it.name == name }.stringField("text")

  @Test
  fun `取出 data 与 time`() {
    val envelope = parseNgaJson("{\"data\":{\"0\":\"ok\"},\"time\":1786111705}")

    assertEquals("ok", envelope.data!!.jsonObject.getValue("0").jsonPrimitive.content)
    assertEquals(1786111705L, envelope.time)
    assertNull(envelope.fakeError)
  }

  @Test
  fun `顶层既没有 data 也没有 error 时默认报解析错`() {
    val error = assertFailsWith<NgaError> {
      parseNgaJson("{\"code\":0,\"msg\":\"\",\"result\":[]}", "direct")
    }

    assertEquals(NgaErrorKind.PARSE, error.kind)
    assertTrue(error.retryable, "解析错要可重试,链才会换下一个组合")
    assertEquals("direct", error.via)
  }

  @Test
  fun `调用方显式声明 bare 时顶层才当 data`() {
    val envelope = parseNgaJson("{\"code\":0,\"msg\":\"\",\"result\":[]}", null, EnvelopeShape.BARE)

    assertEquals("0", envelope.data!!.jsonObject.getValue("code").jsonPrimitive.content)
  }

  @Test
  fun `有 data 壳的响应不受 bare 影响`() {
    val envelope = parseNgaJson("{\"data\":{\"0\":\"ok\"}}", null, EnvelopeShape.BARE)

    assertEquals("ok", envelope.data!!.jsonObject.getValue("0").jsonPrimitive.content)
  }

  @Test
  fun `真错误抛 server 错误且不重试`() {
    val error = assertFailsWith<NgaError> {
      parseNgaJson(captureText("capture-read-thread-not-found"), "direct")
    }

    assertEquals(NgaErrorKind.SERVER, error.kind)
    assertTrue(error.text.contains("找不到主题"))
    assertEquals(false, error.retryable)
    assertEquals("direct", error.via)
  }

  @Test
  fun `假错误当成功返回,但把错误信息留给调用方判空`() {
    val envelope = parseNgaJson(captureText("capture-ucp-not-found"))

    assertEquals("找不到用户", envelope.fakeError?.message)
    assertNull(envelope.data, "只有 error 的响应,data 必须是「没有」而不是空对象")
  }

  @Test
  fun `解析失败抛 parse 错误且可重试`() {
    val error = assertFailsWith<NgaError> { parseNgaJson("<html>你被封了</html>") }

    assertEquals(NgaErrorKind.PARSE, error.kind)
    assertTrue(error.retryable)
    assertTrue(error.text.startsWith("响应不是合法 JSON"), "实际是:${error.text}")
  }

  @Test
  fun `顶层是合法 JSON 但不是对象时报的是另一句话`() {
    val error = assertFailsWith<NgaError> { parseNgaJson("[1,2,3]") }

    assertEquals(NgaErrorKind.PARSE, error.kind)
    assertEquals("响应顶层不是对象", error.text)
  }

  @Test
  fun `空响应也算 parse 错误`() {
    val error = assertFailsWith<NgaError> { parseNgaJson("") }

    assertEquals(NgaErrorKind.PARSE, error.kind)
    assertEquals("响应为空", error.text)
  }

  @Test
  fun `真实抓包 · 通知接口拿到 data`() {
    val envelope = parseNgaJson(captureText("capture-noti-empty"))

    assertEquals("", envelope.data!!.jsonObject.getValue("0").jsonPrimitive.content)
  }

  @Test
  fun `真实抓包 · 用户资料`() {
    val user = parseNgaJson(captureText("capture-ucp-user")).data!!.jsonObject.getValue("0").jsonObject

    assertEquals("41417929", user.getValue("uid").jsonPrimitive.content)
    assertEquals("BugenZhao", user.getValue("username").jsonPrimitive.content)
  }

  @Test
  fun `真实抓包 · 主题列表`() {
    val data = parseNgaJson(captureText("capture-thread-list")).data!!.jsonObject

    assertEquals("原神", data.getValue("__F").jsonObject.getValue("name").jsonPrimitive.content)
    assertTrue(data.getValue("__T").jsonObject.size > 10)
  }

  @Test
  fun `真实抓包 · read_php 的 lite=js 前缀被剥掉后能解析`() {
    val data = parseNgaJson(captureText("capture-read-thread-jsvar")).data!!.jsonObject

    assertNotNull(data["__R"])
  }

  @Test
  fun `坏字节的响应抛 parse 而不是 server`() {
    val error = assertFailsWith<NgaError> {
      parseNgaJson(captureText("capture-thread-list-414-broken-bytes"), "direct")
    }

    assertEquals(NgaErrorKind.PARSE, error.kind)
    assertTrue(error.retryable)
  }
}
